File: Harvard_CS50_Introduction_to_CS/CS50-Pset2-master/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>
#include <string.h>
#include <ctype.h>
int main(void)
{
    int length = 0;
    long card =0;
    long a = 0;
    long powerNum(int a, int b);
    int luhnAlgorithm1(int c, long n, int a);
    int luhnAlgorithm2(int c, long n, int a);
    while (!(card >0))
    {
        card = get_long("Number: ");
        a = card;
    }
    while (a > 0)
    {
        a = a/10;
        length +=1;
    }
    int sum = luhnAlgorithm1(length, card , length) + luhnAlgorithm2(length, card , length);
    if ((length == 13 || length == 15  || length == 16) && ((sum % 10) == 0))
    {
        if (length == 15)
        {
            if (card >= powerNum(34,15) && card < powerNum(35,15))
            {
                printf("AMEX\n");
            }
            else if (card >= powerNum(37,15) && card < powerNum(38,15))
            {
                printf("AMEX\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            if (card >= powerNum(40,13) && card < powerNum(50,13))
            {
                printf("VISA\n");
            }
            else if (card >= powerNum(40,16) && card < powerNum(50,16))
            {
                printf("VISA\n");
            }
            else if (card >= powerNum(51,16) && card < powerNum(56,16))
            {
                printf("MASTERCARD\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

    //function
long powerNum(int a, int b)
{
    return a*pow(10,b-2);
}

int luhnAlgorithm1(int c, long n, int a)
{
    if (c < 2)
    {
        return 0;
    }
    else
    {
        int x = 2*(int)((n % (long)pow(10,a-c+2)) / pow(10, a-c+1));
        if (x < 10)
        {
            return x + luhnAlgorithm1(c-2, n, a);
        }
        else
        {
            return 1 + (int)(x%10) + luhnAlgorithm1(c-2, n, a);
        }
    }
}

int luhnAlgorithm2(int c, long n, int a)
{
    if (c < 1)
    {
        return 0;
    }
    else
    {
        return ((n % (long)pow(10,a-c+1)) / pow(10, a-c)) + luhnAlgorithm2(c-2, n, a);
    }
}

File: CS50_2020/Week6/Credit/credit.py
# import get_int from cs50
from cs50 import get_int

# get user's input
n = 0
while (n <= 0):
    n = get_int("Number: ")
# convert to string to make the char calling easier
s = str(n)
length = len(s)
# sum degits
sum_other = 0
sum_those = 0
for i in range(length - 2, -1, -2):
    temp = 2 * int(s[i])
    sum_other += temp // 10 + temp % 10
for i in range(length - 1, -1, -2):
    sum_those += int(s[i])
add_sum = sum_those + sum_other
# check if sum%10 = 0?
if (add_sum % 10) != 0:
    print("INVALID")
else:
    # take 2 first number
    start_nums = int(s[0]) * 10 + int(s[1])
    # check if the card is a real card with 13 or 15 or 16 numbers and start with valid digits
    if (length == 15 and (start_nums == 34 or start_nums == 37)):
        print("AMEX")
    elif (length == 16 and (start_nums >= 51 and start_nums <= 55)):
        print("MASTERCARD")
    elif ((length == 13 or length == 16) and start_nums >= 40 and start_nums < 50):
        print("VISA")
    else:
        print("INVALID")
